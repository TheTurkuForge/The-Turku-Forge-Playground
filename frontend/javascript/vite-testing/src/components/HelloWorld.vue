<!-- HelloWorld.vue -->
<template>
  <h1>{{ msg }}</h1>
  <button @click="count++">count is: {{ count }}</button>
  <Links />
</template>

<script lang="ts">
import { ref, defineComponent } from "vue";
import Links from "@/components/Links.vue";

export default defineComponent({
  name: "HelloWorld",
  components: {
    Links,
  },
  props: {
    msg: {
      type: String,
      required: true,
    },
  },
  setup: () => {
    const count = ref(0);
    return { count };
  },
});
</script>
