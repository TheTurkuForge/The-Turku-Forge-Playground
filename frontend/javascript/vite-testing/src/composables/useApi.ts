// useApi.ts
interface UserData {
  age: number;
  name: string;
}

const API_RESPONSE_DELAY = 1000; // ms

export const useApi = async () => {
  const userData: UserData = await new Promise((resolve) => {
    const response = { age: 30, name: "John Doe" };
    setTimeout(() => resolve(response), API_RESPONSE_DELAY);
  });

  return {
    userData,
  };
};
