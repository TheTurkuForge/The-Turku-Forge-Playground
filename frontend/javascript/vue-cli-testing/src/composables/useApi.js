export const useApi = async () => {
  await new Promise((r) => setTimeout(r, 1000));
  const api = { getPromo: () => [{ name: "promo1" }] };
  console.log("useApi");
  return {
    api,
  };
};
