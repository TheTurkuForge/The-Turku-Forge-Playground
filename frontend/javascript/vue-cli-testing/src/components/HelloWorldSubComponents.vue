<template>
  <div class="hello">
    <pre>{{ api.getPromo }}</pre>
    <h1>{{ msg }}</h1>
    <p>
      For a guide and recipes on how to configure / customize this project,<br />
      check out the
      <a href="https://cli.vuejs.org" target="_blank" rel="noopener"
      >vue-cli documentation</a
      >.
    </p>
    <CLIPlugins />
    <EssentialLinks />
    <Ecosystem />
  </div>
</template>

<script>
import { useApi } from "@/composables/useApi";
import CLIPlugins from "@/components/CLIPlugins";
import EssentialLinks from "@/components/EssentialLinks";
import Ecosystem from "@/components/Ecosystem";

export default {
  name: "HelloWorldSubComponents",
  components: {Ecosystem, EssentialLinks, CLIPlugins},
  props: {
    msg: String,
  },
  async setup() {
    const api = await useApi();
    return api;
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
</style>
