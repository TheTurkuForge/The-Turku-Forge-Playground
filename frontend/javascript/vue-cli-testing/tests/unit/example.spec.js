import HelloWorld from "@/components/HelloWorld.vue";
import HelloWorldSubComponents from "@/components/HelloWorldSubComponents";
import { useApi } from "@/composables/useApi";
import { flushPromises } from "@vue/test-utils";
import { mountSuspense } from "./mountSuspense";
import suspenseShallowMount, { FALLBACK_TEXT } from "./SuspenseShallowMount";

jest.mock("@/composables/useApi", () => ({
  useApi: jest.fn(),
}));

describe("Home.vue", () => {
  useApi.mockResolvedValue({
    api: {
      getPromo: jest.fn().mockResolvedValue([{ name: "promo1" }]),
    },
  });

  it("renders props.msg with `mountSuspense`", async () => {
    const msg = "new message";
    const props = { msg };
    const wrapper = mountSuspense(HelloWorld, props);

    await flushPromises();

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.text()).toMatch(msg);
  });

  it("renders props.msg with `mountSuspense` and stubs out all other components", async () => {
    const msg = "new message";
    const props = { msg };
    const stubs = Object.keys(HelloWorldSubComponents.components);
    const options = {
      global: {
        stubs,
      },
    };
    const wrapper = mountSuspense(HelloWorldSubComponents, props, options);

    await flushPromises();

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.text()).toMatch(msg);
  });

  it("renders props.msg", async () => {
    const msg = "new message";
    const wrapper = suspenseShallowMount(HelloWorld, { props: { msg } });

    expect(wrapper.text()).toMatch(FALLBACK_TEXT);
    expect(wrapper.element).toMatchSnapshot();

    await flushPromises();

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.text()).toMatch(msg);
  });

  it("renders props.msg and stubs out all other components", async () => {
    const msg = "new message";
    const wrapper = suspenseShallowMount(HelloWorldSubComponents, {
      props: { msg },
    });

    expect(wrapper.text()).toMatch(FALLBACK_TEXT);
    expect(wrapper.element).toMatchSnapshot();

    await flushPromises();

    expect(wrapper.text()).toMatch(msg);
    expect(wrapper.element).toMatchSnapshot();
  });
});
