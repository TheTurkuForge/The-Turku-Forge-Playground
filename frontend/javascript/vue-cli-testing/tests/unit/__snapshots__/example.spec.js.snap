// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue renders props.msg 1`] = `
<span>
  Loading...
</span>
`;

exports[`Home.vue renders props.msg 2`] = `
<div
  class="hello"
>
  <pre>
    function mockConstructor() {
        return fn.apply(this, arguments);
      }
  </pre>
  <h1>
    new message
  </h1>
  <p>
     For a guide and recipes on how to configure / customize this project,
    <br />
     check out the 
    <a
      href="https://cli.vuejs.org"
      rel="noopener"
      target="_blank"
    >
      vue-cli documentation
    </a>
    . 
  </p>
  <h3>
    Installed CLI Plugins
  </h3>
  <ul>
    <li>
      <a
        href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel"
        rel="noopener"
        target="_blank"
      >
        babel
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-router"
        rel="noopener"
        target="_blank"
      >
        router
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest"
        rel="noopener"
        target="_blank"
      >
        unit-jest
      </a>
    </li>
  </ul>
  <h3>
    Essential Links
  </h3>
  <ul>
    <li>
      <a
        href="https://vuejs.org"
        rel="noopener"
        target="_blank"
      >
        Core Docs
      </a>
    </li>
    <li>
      <a
        href="https://forum.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        Forum
      </a>
    </li>
    <li>
      <a
        href="https://chat.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        Community Chat
      </a>
    </li>
    <li>
      <a
        href="https://twitter.com/vuejs"
        rel="noopener"
        target="_blank"
      >
        Twitter
      </a>
    </li>
    <li>
      <a
        href="https://news.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        News
      </a>
    </li>
  </ul>
  <h3>
    Ecosystem
  </h3>
  <ul>
    <li>
      <a
        href="https://router.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        vue-router
      </a>
    </li>
    <li>
      <a
        href="https://vuex.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        vuex
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/vue-devtools#vue-devtools"
        rel="noopener"
        target="_blank"
      >
        vue-devtools
      </a>
    </li>
    <li>
      <a
        href="https://vue-loader.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        vue-loader
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/awesome-vue"
        rel="noopener"
        target="_blank"
      >
        awesome-vue
      </a>
    </li>
  </ul>
</div>
`;

exports[`Home.vue renders props.msg and stubs out all other components 1`] = `
<span>
  Loading...
</span>
`;

exports[`Home.vue renders props.msg and stubs out all other components 2`] = `
<div
  class="hello"
>
  <pre>
    function mockConstructor() {
        return fn.apply(this, arguments);
      }
  </pre>
  <h1>
    new message
  </h1>
  <p>
     For a guide and recipes on how to configure / customize this project,
    <br />
     check out the 
    <a
      href="https://cli.vuejs.org"
      rel="noopener"
      target="_blank"
    >
      vue-cli documentation
    </a>
    . 
  </p>
  <c-l-i-plugins-stub />
  <essential-links-stub />
  <ecosystem-stub />
</div>
`;

exports[`Home.vue renders props.msg with \`mountSuspense\` 1`] = `
<div
  class="hello"
>
  <pre>
    function mockConstructor() {
        return fn.apply(this, arguments);
      }
  </pre>
  <h1>
    new message
  </h1>
  <p>
     For a guide and recipes on how to configure / customize this project,
    <br />
     check out the 
    <a
      href="https://cli.vuejs.org"
      rel="noopener"
      target="_blank"
    >
      vue-cli documentation
    </a>
    . 
  </p>
  <h3>
    Installed CLI Plugins
  </h3>
  <ul>
    <li>
      <a
        href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel"
        rel="noopener"
        target="_blank"
      >
        babel
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-router"
        rel="noopener"
        target="_blank"
      >
        router
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-jest"
        rel="noopener"
        target="_blank"
      >
        unit-jest
      </a>
    </li>
  </ul>
  <h3>
    Essential Links
  </h3>
  <ul>
    <li>
      <a
        href="https://vuejs.org"
        rel="noopener"
        target="_blank"
      >
        Core Docs
      </a>
    </li>
    <li>
      <a
        href="https://forum.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        Forum
      </a>
    </li>
    <li>
      <a
        href="https://chat.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        Community Chat
      </a>
    </li>
    <li>
      <a
        href="https://twitter.com/vuejs"
        rel="noopener"
        target="_blank"
      >
        Twitter
      </a>
    </li>
    <li>
      <a
        href="https://news.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        News
      </a>
    </li>
  </ul>
  <h3>
    Ecosystem
  </h3>
  <ul>
    <li>
      <a
        href="https://router.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        vue-router
      </a>
    </li>
    <li>
      <a
        href="https://vuex.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        vuex
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/vue-devtools#vue-devtools"
        rel="noopener"
        target="_blank"
      >
        vue-devtools
      </a>
    </li>
    <li>
      <a
        href="https://vue-loader.vuejs.org"
        rel="noopener"
        target="_blank"
      >
        vue-loader
      </a>
    </li>
    <li>
      <a
        href="https://github.com/vuejs/awesome-vue"
        rel="noopener"
        target="_blank"
      >
        awesome-vue
      </a>
    </li>
  </ul>
</div>
`;

exports[`Home.vue renders props.msg with \`mountSuspense\` and stubs out all other components 1`] = `
<div
  class="hello"
>
  <pre>
    function mockConstructor() {
        return fn.apply(this, arguments);
      }
  </pre>
  <h1>
    new message
  </h1>
  <p>
     For a guide and recipes on how to configure / customize this project,
    <br />
     check out the 
    <a
      href="https://cli.vuejs.org"
      rel="noopener"
      target="_blank"
    >
      vue-cli documentation
    </a>
    . 
  </p>
  <c-l-i-plugins-stub />
  <essential-links-stub />
  <ecosystem-stub />
</div>
`;
